pkg/controller/installation/windows/calico_windows_upgrader.go

```diff
@@ -0,0 +1,456 @@
+			return nil, nil, nil, fmt.Errorf("Never expected index to have anything other than a Node object: %v", obj)
+		}
+
+		if node.Labels[corev1.LabelOSStable] != "windows" {
+			continue
+		}
+
+		windowsLog.V(1).Info(fmt.Sprintf("Processing node %v", node.Name))
+		exists, variant, version := common.GetNodeVariantAndVersion(node)
+
+		// If the version annotation doesn't exist, something is wrong with the
+		// Calico Windows node or it needs to be upgraded manually to
+		// a version supported by the calicoWindowsUpgrader.
+		if !exists {
+			return nil, nil, nil, fmt.Errorf("Node %v does not have the version annotation, it might be unhealthy or it might be running an unsupported Calico version.", node.Name)
+		}
+
+		if node.Labels[common.CalicoWindowsUpgradeLabel] == common.CalicoWindowsUpgradeLabelInProgress {
+			windowsLog.V(1).Info(fmt.Sprintf("Node %v has the upgrade in-progress label", node.Name))
+			inProgress[node.Name] = node
+		} else if variant != w.install.Variant || version != expectedVersion {
+			windowsLog.V(1).Info(fmt.Sprintf("Node %v doesn't have the latest variant and/or version. variant=%v, expectedVariant=%v, version=%v, expectedVersion=%v", node.Name, variant, w.install.Variant, version, expectedVersion))
+			pending[node.Name] = node
+		} else {
+			windowsLog.V(1).Info(fmt.Sprintf("Node %v has the latest variant and version", node.Name))
+			inSync[node.Name] = node
+		}
+	}
+
+	windowsLog.V(1).Info(fmt.Sprintf("pending=%v, in-progress=%v, in-sync=%v", len(pending), len(inProgress), len(inSync)))
+	return pending, inProgress, inSync, nil
+}
+
+func (w *calicoWindowsUpgrader) upgradeCompleted(node *corev1.Node) bool {
+	_, variant, version := common.GetNodeVariantAndVersion(node)
+	if variant == w.install.Variant && version == w.getExpectedVersion() {
+		return true
+	}
+	return false
+}
+
+func (w *calicoWindowsUpgrader) isUpgradeFromCalicoToEnterprise(node *corev1.Node) bool {
+	_, variant, _ := common.GetNodeVariantAndVersion(node)
+	if variant == operatorv1.Calico && w.install.Variant == operatorv1.TigeraSecureEnterprise {
+		return true
+	}
+	return false
+}
+
+func (w *calicoWindowsUpgrader) getExpectedVersion() string {
+	if w.install.Variant == operatorv1.TigeraSecureEnterprise {
+		return components.EnterpriseRelease
+	}
+	return components.CalicoRelease
+}
+
+func sortedSliceFromMap(m map[string]*corev1.Node) []string {
+	nodeNames := make([]string, 0, len(m))
+	// Copy map keys to slice and sort.
+	for nodeName, _ := range m {
+		nodeNames = append(nodeNames, nodeName)
+	}
+	sort.Strings(nodeNames)
+	return nodeNames
+}
+
+func (w *calicoWindowsUpgrader) updateWindowsNodes() {
+	pending, inProgress, inSync, err := w.getNodeUpgradeStatus()
+
+	// Make sorted slices of the node names from the node upgrade status maps.
+	// We'll loop through these slices to process node upgrades
+	// deterministically.
+	pendingNodeNames := sortedSliceFromMap(pending)
+	inProgressNodeNames := sortedSliceFromMap(inProgress)
```

Comment:
You need to move this logic to just before where we use it - the sorted slice isn't going to be accurate, because we modify the maps as part of the loop below.

For example, we move entries from in-progress to in-sync in the first loop.

Reply:
I'd probably even just recommend doing it like this:

```go
for _, nodeName := range sortedSliceFromMap(inProgress) {
}
```

Reply:
fixed in 8dad23754491bd4fa5e4513d2b7d06becb7e2f0a